flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlCreateTable.java

```diff
@@ -310,9 +322,67 @@ private void printIndent(SqlWriter writer) {
 		public List<SqlNode> columnList = new ArrayList<>();
 		public SqlNodeList primaryKeyList = SqlNodeList.EMPTY;
 		public List<SqlNodeList> uniqueKeysList = new ArrayList<>();
+		@Nullable public SqlWatermark watermark;
 	}
 
 	public String[] fullTableName() {
 		return tableName.names.toArray(new String[0]);
 	}
+
+	// -------------------------------------------------------------------------------------
+
+	private static final class ColumnValidator {
+
+		private final Set<String> allColumnNames = new HashSet<>();
+
+		/**
+		 * Adds column name to the registered column set. This will add nested column names recursive.
+		 * Nested column names are qualified using "." separator.
+		 */
+		public void addColumn(SqlNode column) throws SqlValidateException {
+			String columnName;
+			if (column instanceof SqlTableColumn) {
+				SqlTableColumn tableColumn = (SqlTableColumn) column;
+				columnName = tableColumn.getName().getSimple();
+				addNestedColumn(columnName, tableColumn.getType());
+			} else if (column instanceof SqlBasicCall) {
+				SqlBasicCall tableColumn = (SqlBasicCall) column;
+				columnName = tableColumn.getOperands()[1].toString();
+			} else {
+				throw new UnsupportedOperationException("Unsupported column:" + column);
+			}
+
+			addColumnName(columnName, column.getParserPosition());
+		}
+
+		/**
+		 * Returns true if the column name is existed in the registered column set.
+		 * This supports qualified column name using "." separator.
+		 */
+		public boolean contains(String columnName) {
+			return allColumnNames.contains(columnName);
+		}
+
+		private void addNestedColumn(String columnName, SqlDataTypeSpec columnType) throws SqlValidateException {
+			SqlTypeNameSpec typeName = columnType.getTypeNameSpec();
+			// validate composite type
+			if (typeName instanceof ExtendedSqlRowTypeNameSpec) {
```

Comment:
Do we support nested collection type such as `ARRAY` or `MULTISET` ?

Reply:
No. We don't support it.

Reply:
If that is the case, we should match the `ARRAY` and `MULTISET` type explicitly in the `ColumnValidator`, and throws a message that is readable.

Reply:
What I mean "not support" is rowtime field should not be a `array[1]` or `map['key1']` , and this is guaranteed by sql parser for indentifer.  We can of course allow `ARRAY` and `MULTISET` as the nested type.

Reply:
Why do you think a row time field "should not be" a `array[1]` or `map['key1']` ?

Reply:
This is not allowed by the design of FLIP-66.

Reply:
`array[1]` or `map['key1']` are expressions, not field reference. The watermark statement design aims to have an **existing column** as the rowtime field. If it doesn't exist in the schema, we can use computed column to derive it.

Reply:
Then why shall we allow nested field for structure type ? I didn't see the difference.

Reply:
@twalthr , what do you think about this?

Reply:
I think this is a design choice. The motivation is we want to make sure the event time column exists when defining watermark. So we lean to only support the types which we can get enough information to do the static check. `ROW` and other basic types are allowed to do static check but not for collection types such as `ARRAY` and `MAP`, since you never know whether if `array[1]` or  `map['key1']` really exist or not. 

It's just a safe choice, nothing else.

Reply:
@danny0405 Does this resolve your concern?

Reply:
For "Does not exist", do you mean the value is null ? A record type field can also have null value.

BTW, the user did knows that if the `array[1]` exists, i don't think we forbidden the syntax just for a "safe" reason. Actually, in the standard SQL, `ARRAY` type has a fixed length, that means, its structure is fixed.

For `MAP` type, i think it's equivalent with a Java POJO or record, we support the record but not the map, that does not make sense to me.

Reply:
I think the main reason is implementation level. In the implementation, we recognize a time attribute field by a specical data type `TimeIndicatorType`. But for a `ArrayType<elementType>`, we can't say the 3rd element in the array is `TimeIndicatorType`, all the elements in the array share the same data type `elementType`. That's why we only support field reference as time attribute field, because we can recognize them by datat type.

Reply:
Well, exactly, because of the implementation limitation, supporting only nested Record type seems reasonable.